kv: fix cleanupTxn race

RaduBerinde opened the pull request:
There is a window between when `cleanupTxn` runs and when the heartbeat
goroutine unregisters the txn. During this window another op might start on the
txn, causing a second call to `cleanupTxn`, which crashes (close of nil
channel). This change fixes this by adding a check for the nil channel.

TestTxnCoordIdempotentCleanup is resurrected, though it had to be rewritten.
Verified that the test causes the crash without the fix.  I also ran all tests
with a 1ms sleep added in heartbeatLoop before unregistering the txn.

Fixes #4426.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5408)

<!-- Reviewable:end -->

bdarnell commented:
It makes me nervous to allow `cleanupTxn` to be called multiple times (for example, the second call's `Transaction` argument will be ignored instead of getting assigned to `txnMeta.txn`. That may be fine in this case, but it doesn't feel right). What if instead we instead extended the check in `TxnCoordSender.Send` (line 344) to return the "writing transaction timed out, was aborted, or ran on multiple coordinators" error if `txnMeta.txnEnd` is nil?

RaduBerinde commented:
That check only happens if the txn is Writing. So are you proposing that we check txnMeta in all cases, and exit if `ba.Txn.Writing && !ok` or `ok && txnMeta.txnEnd`? That should work AFAICT. What do you think @tschottdorf?

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5408#-:-KDK3po0VXpnnZLrKEvI:-1537061526)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I think that's a good idea, but we don't have a lock on the txnMeta itself for long, so I think it would be racy (though I haven't checked). Two concurrent calls could sneak through and still double-finish.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5408#-:-KDK7cknkwZSuoYi17Fj:631172824)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Right, if we have concurrent calls it can still happen. I am not sure if (or how thoroughly) we are testing this layer with concurrent calls (the higher layers probably don't do that, at least not in normal circumstances).

Anyway, @bdarnell @tschottdorf I will let you two decide since my knowledge of this layer is very limited.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5408#-:-KDPByLhZElCmMIWjeVP:-456217386)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
The other issue is that multiple EndTransaction are legal (for example, on txn restarts; what's not legal is committing/aborting and then doing more stuff with the transaction). So really what we want to track is whether there's an outstanding `EndTransactionRequest`... seems more trouble than it's worth. I'm happy to keep the current version, add a TODO (or file an issue) and then revisit post-beta when we're retrofitting this code path to handle parallelism better.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5408#-:-KDPOetaUcmd1mycPB0E:-921670347)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
What do outstanding `EndTransactionRequests` have to do with it? For retries, there may be multiple EndTransaction calls, but only the final one will return a non-pending status and trigger a `cleanupTxn`.

Anyway, looking closer at this it looks like it will be tricky to ensure that we call `cleanupTxn` exactly once - we don't hold locks with the appropriate scope.  So this version LGTM

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5408#-:-KDPRy8VBK97xSrePPZT:-1065202034)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> What do outstanding EndTransactionRequests have to do with it?

I was thinking about nil'ing out the channel early the first time we get the lock, but you only know you want to do that when the `EndTransaction` comes back. Should've mentioned that.

RaduBerinde committed:
kv: fix cleanupTxn race

There is a window between when `cleanupTxn` runs and when the heartbeat
goroutine unregisters the txn. During this window another op might start on the
txn, causing a second call to `cleanupTxn`, which crashes (close of nil
channel). This change fixes this by adding a check for the nil channel.

TestTxnCoordIdempotentCleanup is resurrected, though it had to be rewritten.
Verified that the test causes the crash without the fix. Ran it with make
stress. I also ran all tests with a 1ms sleep added in heartbeatLoop before
unregistering the txn.

Fixes #4426.
